In [1]:
import numpy as np, statsmodels.stats.api as sms
import itertools
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.ticker import NullFormatter
import pandas as pd
import numpy as np
import matplotlib.ticker as ticker
import random
from sklearn import preprocessing
import math

In [2]:
data=pd.read_csv('project_data.csv')
data.head()

user                              adid week  expid  if_click  \
0  297353  0b0fa14b56d3741178196daaa92e6a1e  w_2      1         0   
1   98719  ea875acb76e0a806a7837174528f62d9  w_3      1         0   
2  205144  1f1fe825014d9e9a0881233d9950bd43  w_2      1         0   
3  559194  1f1fe825014d9e9a0881233d9950bd43  w_2      1         0   
4   11888  0d82161270febc99f7defb653e339113  w_3      0         0   

   real_like_cnt    category  brand_effect  experience  status  user_age  \
0              3         Car             1           1       1        48   
1              3         Car             1           1       1        29   
2              1     Jewelry             0           1       1        24   
3              5     Jewelry             0           1       1        26   
4              1  Cosmetrics             0           1       1        40   

   friend_age user_gender friend_gender user_city friend_city  user_degree  \
0          37      female          male   level_1     level_1          495   
1          34      female        female   level_3     level_2          479   
2          24      female          male   level_2     level_1          682   
3          41        male          male   level_1     level_1          340   
4          34      female        female   level_1     level_2           95   

   friend_degree  user_sns_like_cnt  user_sns_comment_cnt  
0            436                162                   138  
1           3911                212                   312  
2            470                 95                   336  
3            602                 35                   138  
4            836                 50                    38

# For whole sample

In [3]:
#find the number of control and treatment group
d_0 = data[data['expid'] == 0]['if_click']
d_1 = data[data['expid'] == 1]['if_click']

In [4]:
print(len(d_1),len(d_0))

500000 500000


#### The sample size of control group is 500,000 and is equal to the sample size of treatment group  

#### Assume we want to detect 10% lift about click_rate after treatment

In [5]:
#to get the variance
lift=0.1  #10%
p=np.mean(d_0)
var=p*(1-p)
#to get the standard error
std=math.sqrt(var)
# the difference we want to detect
mean_difference=lift*p

In [6]:
from statsmodels.stats.power import zt_ind_solve_power

In [7]:
size_80=zt_ind_solve_power(effect_size=(mean_difference/std), alpha=0.05, power=0.8, ratio=1,alternative="two-sided")
size_95=zt_ind_solve_power(effect_size=(mean_difference/std), alpha=0.05, power=0.95, ratio=1,alternative="two-sided")
size_99=zt_ind_solve_power(effect_size=(mean_difference/std), alpha=0.05, power=0.99, ratio=1,alternative="two-sided")

In [8]:
print(size_80,size_95,size_99)

21998.964038254322 36421.86502357748 51494.77309889534


#### To acheive 99% Power Analysis with 5%α, the experiment need sample size which is greater than 51494. However, the true sample sizes of control group and treatment group are both 500,000, which means this experiment has acheived 99% Power Analysis, and is hard to committee Type II error.

# For different category

In [9]:
#find the number of control and treatment group for each type of ads category
data[data['expid'] == 0]['category'].value_counts()

Cosmetrics    100000
Car           100000
Jewelry       100000
Clothes       100000
Beverage      100000
Name: category, dtype: int64

In [10]:
data[data['expid'] == 1]['category'].value_counts()

Car           100000
Jewelry       100000
Cosmetrics    100000
Beverage      100000
Clothes       100000
Name: category, dtype: int64

## Car

In [11]:
dc=data[data['category']=='Car']

In [12]:
dc_0=dc[dc['expid'] == 0]['if_click']
dc_1=dc[dc['expid'] == 1]['if_click']

In [13]:
print(len(dc_0))
print(len(dc_1))

100000
100000


In [14]:
#to get the variance
p_c=np.mean(dc_0)
var_c=p_c*(1-p_c)
#to get the standard error
std_c=math.sqrt(var_c)
# the difference we want to detect
mean_difference_c=lift*p_c

In [15]:
size_80_c=zt_ind_solve_power(effect_size=(mean_difference_c/std_c), alpha=0.05, power=0.8, ratio=1,alternative="two-sided")
size_95_c=zt_ind_solve_power(effect_size=(mean_difference_c/std_c), alpha=0.05, power=0.95, ratio=1,alternative="two-sided")
size_99_c=zt_ind_solve_power(effect_size=(mean_difference_c/std_c), alpha=0.05, power=0.99, ratio=1,alternative="two-sided")

In [16]:
print(size_80_c,size_95_c,size_99_c)

22506.48712919673 37262.12902367005 52682.77387818322


#### The sample sizes of control group and treatment group for car ads are both 100,000, greater than 52682, which means it has acheived 99% Power Analysis, and is hard to committee Type II error.

## Cosmetrics

In [17]:
dC=data[data['category']=='Cosmetrics']
dC_0=dC[dC['expid'] == 0]['if_click']
dC_1=dC[dC['expid'] == 1]['if_click']
print(len(dC_0))
print(len(dC_1))

100000
100000


In [18]:
#to get the variance
p_C=np.mean(dC_0)
var_C=p_C*(1-p_C)
#to get the standard error
std_C=math.sqrt(var_C)
# the difference we want to detect
mean_difference_C=lift*p_C

In [19]:
size_80_C=zt_ind_solve_power(effect_size=(mean_difference_C/std_C), alpha=0.05, power=0.8, ratio=1,alternative="two-sided")
size_80_C=zt_ind_solve_power(effect_size=(mean_difference_C/std_C), alpha=0.05, power=0.8, ratio=1,alternative="two-sided")
size_95_C=zt_ind_solve_power(effect_size=(mean_difference_C/std_C), alpha=0.05, power=0.95, ratio=1,alternative="two-sided")
size_99_C=zt_ind_solve_power(effect_size=(mean_difference_C/std_C), alpha=0.05, power=0.99, ratio=1,alternative="two-sided")

In [20]:
print(size_80_C,size_95_C,size_99_C)

21610.56017598977 35778.81687773284 50585.6044349504


#### The sample sizes of control group and treatment group for Cosmetrics ads are both 100,000, greater than 50585, which means it has acheived 99% Power Analysis, and is hard to committee Type II error.

## Jewelry

In [21]:
dC=data[data['category']=='Jewelry']
dC_0=dC[dC['expid'] == 0]['if_click']
dC_1=dC[dC['expid'] == 1]['if_click']
print(len(dC_0))
print(len(dC_1))

100000
100000


In [22]:
#to get the variance
p_C=np.mean(dC_0)
var_C=p_C*(1-p_C)
#to get the standard error
std_C=math.sqrt(var_C)
# the difference we want to detect
mean_difference_C=lift*p_C

In [23]:
size_80_C=zt_ind_solve_power(effect_size=(mean_difference_C/std_C), alpha=0.05, power=0.8, ratio=1,alternative="two-sided")
size_80_C=zt_ind_solve_power(effect_size=(mean_difference_C/std_C), alpha=0.05, power=0.8, ratio=1,alternative="two-sided")
size_95_C=zt_ind_solve_power(effect_size=(mean_difference_C/std_C), alpha=0.05, power=0.95, ratio=1,alternative="two-sided")
size_99_C=zt_ind_solve_power(effect_size=(mean_difference_C/std_C), alpha=0.05, power=0.99, ratio=1,alternative="two-sided")

In [24]:
print(size_80_C,size_95_C,size_99_C)

19142.354850574808 31692.413491548996 44808.07450741644


#### The sample sizes of control group and treatment group for Jewelry ads are both 100,000, greater than 44808, which means it has acheived 99% Power Analysis, and is hard to committee Type II error.

## Clothes

In [25]:
dC=data[data['category']=='Clothes']
dC_0=dC[dC['expid'] == 0]['if_click']
dC_1=dC[dC['expid'] == 1]['if_click']
print(len(dC_0))
print(len(dC_1))

100000
100000


In [26]:
#to get the variance
p_C=np.mean(dC_0)
var_C=p_C*(1-p_C)
#to get the standard error
std_C=math.sqrt(var_C)
# the difference we want to detect
mean_difference_C=lift*p_C

In [27]:
size_80_C=zt_ind_solve_power(effect_size=(mean_difference_C/std_C), alpha=0.05, power=0.8, ratio=1,alternative="two-sided")
size_80_C=zt_ind_solve_power(effect_size=(mean_difference_C/std_C), alpha=0.05, power=0.8, ratio=1,alternative="two-sided")
size_95_C=zt_ind_solve_power(effect_size=(mean_difference_C/std_C), alpha=0.05, power=0.95, ratio=1,alternative="two-sided")
size_99_C=zt_ind_solve_power(effect_size=(mean_difference_C/std_C), alpha=0.05, power=0.99, ratio=1,alternative="two-sided")
print(size_80_C,size_95_C,size_99_C)

24549.564035766565 40644.682457767885 57465.1709736955


#### The sample sizes of control group and treatment group for Clothes ads are both 100,000, greater than 57465, which means it has acheived 99% Power Analysis, and is hard to committee Type II error.

## Beverage

In [28]:
dC=data[data['category']=='Beverage']
dC_0=dC[dC['expid'] == 0]['if_click']
dC_1=dC[dC['expid'] == 1]['if_click']
print(len(dC_0))
print(len(dC_1))

100000
100000


In [29]:
# to get the true difference of click rate
mean_difference_C=np.mean(dC_1) - np.mean(dC_0)
mean_difference_C
#to get the variance
p_C=np.mean(dC_0)
var_C=p_C*(1-p_C)
# the difference we want to detect
mean_difference_C=lift*p_C

In [30]:
size_80_C=zt_ind_solve_power(effect_size=(mean_difference_C/std_C), alpha=0.05, power=0.8, ratio=1,alternative="two-sided")
size_95_C=zt_ind_solve_power(effect_size=(mean_difference_C/std_C), alpha=0.05, power=0.95, ratio=1,alternative="two-sided")
size_99_C=zt_ind_solve_power(effect_size=(mean_difference_C/std_C), alpha=0.05, power=0.99, ratio=1,alternative="two-sided")
print(size_80_C,size_95_C,size_99_C)

21507.37404854061 35607.98014176293 50344.0682311455


#### The sample sizes of control group and treatment group for Beverageads are both 100,000, greater than 50344.0682311455, which means it has acheived 99% Power Analysis, and is hard to committee Type II error.